dbms/src/Storages/Transaction/SSTReader.h

```diff
@@ -14,28 +14,110 @@
 
 #pragma once
 
+#include <Common/Logger.h>
 #include <Common/nocopyable.h>
 #include <Storages/Transaction/ProxyFFI.h>
 
 namespace DB
 {
-
-struct SSTReader
+class SSTReader
 {
-    bool remained() const;
-    BaseBuffView key() const;
-    BaseBuffView value() const;
-    void next();
+public:
+    virtual bool remained() const;
+    virtual BaseBuffView keyView() const;
+    virtual BaseBuffView valueView() const;
+    virtual void next();
 
     DISALLOW_COPY_AND_MOVE(SSTReader);
     SSTReader(const TiFlashRaftProxyHelper * proxy_helper_, SSTView view);
-    ~SSTReader();
+    // Should only be used when SSTReader is a base class.
+    SSTReader();
+    virtual ~SSTReader();
 
 private:
     const TiFlashRaftProxyHelper * proxy_helper;
     SSTReaderPtr inner;
     ColumnFamilyType type;
+    bool inited;
 };
 
+/// MultiSSTReader helps when there are multiple sst files in a column family.
+/// It is derived from virtual class SSTReader, so it can be holded in a SSTReaderPtr.
+/// It also maintains instance of `R` which is normaly SSTReader(and MockSSTReader in tests),
+/// to read a single sst file.
+/// An `Initer` function is need to create a instance of `R` from a instance of `E`,
+/// which is usually path of the SST file.
+/// We introduce `R` and `E` to make this class testable.
+template <typename R, typename E>
+class MultiSSTReader : public SSTReader
+{
+public:
+    using Initer = std::function<std::unique_ptr<R>(const TiFlashRaftProxyHelper *, E)>;
+
+    DISALLOW_COPY_AND_MOVE(MultiSSTReader);
+
+    bool remained() const override
+    {
+        this->maybe_next_reader();
+        return mono->remained();
+    }
+    BaseBuffView keyView() const override
+    {
+        return mono->keyView();
+    }
+    BaseBuffView valueView() const override
+    {
+        return mono->valueView();
+    }
+    void next() override
+    {
+        this->maybe_next_reader();
```

Comment:
What about next after the read instead of before the read? In this way may be we can eliminate the case where current mono is drained.

Reply:
Seems it will fail if a reader is empty, not knowing whether this will happen.

Reply:
You are right. A usual convention (as a cursor like thing) could be:

- `keyView` and `valueView` is valid after an initial `next()`
- The copied value of `keyView` and `valueView` may be invalidated if `next()` is called.

Reply:
Maybe your suggestion is correct, `reader->remain()` can be interpreted into a "valid" function rather than a "hasNext" function.